# BLG527E - MACHINE LEARNING - FALL 2025-2026
## Term Project (DengAI)

* Sedef Kjamılı - 504251563
* Rabia Şevval Aydın - 504251520

Term Project Checklist (DengAI)

* Step 1: Form team of 2 [DONE]

* Step 2: Join DengAI competition (DrivenData) [DONE]

* Step 3: Download all datasets [DONE]

* Step 4: Load and inspect datasets [DONE]

* Step 5: Identify target (total_cases) [DONE]

* Step 6: Handle missing values [DONE]

* Step 7: Separate data by city (San Juan / Iquitos) [DONE]

* Step 8: Time-based train/validation split [DONE]

* Step 9: Prepare features & target (Model-ready) [DONE]

* Step 10: Train Polynomial Regression model [DONE]

* Step 11: Choose and justify polynomial degree (k) [DONE]

* Step 12 : Evaluate using MAE [DONE]

* Step 13 : Select ML model from literature (RF / XGBoost / LSTM / Ensemble) [DONE]

* Step 14 : Cite related work [DONE]

* Step 15 : Train and tune ML model [DONE]

* Step 16 : Evaluate using MAE [DONE]

* Step 17 : Compare models (Polynomial vs ML) [DONE]

* Step 18 : Analyze feature importance [DONE]

* Step 19 : Identify strongest predictors [DONE]

* Step 20 : Build time-ahead forecasting models [DONE]

* Step 21 : Predict 1, 2, 3+ weeks ahead [DONE]

* Step 22 : Evaluate MAE for each horizon [DONE]

* Step 23 : Select best prediction horizon [DONE]

* Step 24 : Train best model on full data [DONE]

* Step 25 : Generate test predictions [DONE]

* Step 26 : Submit to competition [DONE]

* Record leaderboard score [DONE]

* Write ACM-format report [DONE]

* Add tables, figures, and citations [DONE]

* Prepare 5-minute demo video [DONE]

* Submit report, code, and demo to Ninova [DONE]

In [ ]:
#importing libraries
import pandas as pd
import numpy as np

# Step 4: Load and inspect datasets

In [ ]:
#loading the datasets
X_train = pd.read_csv("dengue_features_train.csv")
y_train = pd.read_csv("dengue_labels_train.csv")
X_test  = pd.read_csv("dengue_features_test.csv")
submission = pd.read_csv("submission_format.csv")

In [ ]:
#inspect the data
X_train.head()

,city,year,weekofyear,week_start_date,ndvi_ne,ndvi_nw,ndvi_se,ndvi_sw,precipitation_amt_mm,reanalysis_air_temp_k,...,reanalysis_precip_amt_kg_per_m2,reanalysis_relative_humidity_percent,reanalysis_sat_precip_amt_mm,reanalysis_specific_humidity_g_per_kg,reanalysis_tdtr_k,station_avg_temp_c,station_diur_temp_rng_c,station_max_temp_c,station_min_temp_c,station_precip_mm
0,sj,1990,18,1990-04-30,0.122600,0.103725,0.198483,0.177617,12.42,297.572857,...,32.00,73.365714,12.42,14.012857,2.628571,25.442857,6.900000,29.4,20.0,16.0
1,sj,1990,19,1990-05-07,0.169900,0.142175,0.162357,0.155486,22.82,298.211429,...,17.94,77.368571,22.82,15.372857,2.371429,26.714286,6.371429,31.7,22.2,8.6
2,sj,1990,20,1990-05-14,0.032250,0.172967,0.157200,0.170843,34.54,298.781429,...,26.10,82.052857,34.54,16.848571,2.300000,26.714286,6.485714,32.2,22.8,41.4
3,sj,1990,21,1990-05-21,0.128633,0.245067,0.227557,0.235886,15.36,298.987143,...,13.90,80.337143,15.36,16.672857,2.428571,27.471429,6.771429,33.3,23.3,4.0
4,sj,1990,22,1990-05-28,0.196200,0.262200,0.251200,0.247340,7.52,299.518571,...,12.20,80.460000,7.52,17.210000,3.014286,28.942857,9.371429,35.0,23.9,5.8


# Step 5: Identify target (total_cases)

In [ ]:
#inspect the labels (targets)
y_train.head()

,city,year,weekofyear,total_cases
0,sj,1990,18,4
1,sj,1990,19,5
2,sj,1990,20,4
3,sj,1990,21,3
4,sj,1990,22,6


In [ ]:
#checking the data size
print("X_train shape:", X_train.shape)
print("y_train shape:", y_train.shape)

X_train shape: (1456, 24)
y_train shape: (1456, 4)


In [ ]:
#merging features and labels
train_df = X_train.merge(
    y_train,
    on=["city", "year", "weekofyear"]
)

train_df.head()

,city,year,weekofyear,week_start_date,ndvi_ne,ndvi_nw,ndvi_se,ndvi_sw,precipitation_amt_mm,reanalysis_air_temp_k,...,reanalysis_relative_humidity_percent,reanalysis_sat_precip_amt_mm,reanalysis_specific_humidity_g_per_kg,reanalysis_tdtr_k,station_avg_temp_c,station_diur_temp_rng_c,station_max_temp_c,station_min_temp_c,station_precip_mm,total_cases
0,sj,1990,18,1990-04-30,0.122600,0.103725,0.198483,0.177617,12.42,297.572857,...,73.365714,12.42,14.012857,2.628571,25.442857,6.900000,29.4,20.0,16.0,4
1,sj,1990,19,1990-05-07,0.169900,0.142175,0.162357,0.155486,22.82,298.211429,...,77.368571,22.82,15.372857,2.371429,26.714286,6.371429,31.7,22.2,8.6,5
2,sj,1990,20,1990-05-14,0.032250,0.172967,0.157200,0.170843,34.54,298.781429,...,82.052857,34.54,16.848571,2.300000,26.714286,6.485714,32.2,22.8,41.4,4
3,sj,1990,21,1990-05-21,0.128633,0.245067,0.227557,0.235886,15.36,298.987143,...,80.337143,15.36,16.672857,2.428571,27.471429,6.771429,33.3,23.3,4.0,3
4,sj,1990,22,1990-05-28,0.196200,0.262200,0.251200,0.247340,7.52,299.518571,...,80.460000,7.52,17.210000,3.014286,28.942857,9.371429,35.0,23.9,5.8,6


# Step 6: Handle missing values

In [ ]:
#checking for missing values
train_df.isna().sum().sort_values(ascending=False).head(10)

#output:
#-Missing values exist (expected in real climate data)
#-Most missing:
#---ndvi_ne (194)
#---ndvi_nw (52)
#Some station measurements also missing

,0
ndvi_ne,194
ndvi_nw,52
station_avg_temp_c,43
station_diur_temp_rng_c,43
ndvi_se,22
ndvi_sw,22
station_precip_mm,22
station_max_temp_c,20
station_min_temp_c,14
precipitation_amt_mm,13


In [ ]:
#HandlİNG missing values - Mean imputation per city
train_df_imputed = train_df.copy()

numeric_cols = train_df_imputed.select_dtypes(include=[np.number]).columns

train_df_imputed[numeric_cols] = (
    train_df_imputed
    .groupby("city")[numeric_cols]
    .transform(lambda x: x.fillna(x.mean()))
)

In [ ]:
#verifying that missing values are gone
train_df_imputed.isna().sum().sort_values(ascending=False).head()

,0
city,0
year,0
weekofyear,0
week_start_date,0
ndvi_ne,0


# Step 7: Separate data by city (San Juan / Iquitos)

In [ ]:
train_sj = train_df_imputed[train_df_imputed["city"] == "sj"].sort_values("week_start_date")
train_iq = train_df_imputed[train_df_imputed["city"] == "iq"].sort_values("week_start_date")

print("San Juan shape:", train_sj.shape)
print("Iquitos shape:", train_iq.shape)

San Juan shape: (936, 25)
Iquitos shape: (520, 25)


# Step 8: Time-based train/validation split

In [ ]:
#First 80% → training
#Last 20% → validation

def time_split(df, split_ratio=0.8):
    split_idx = int(len(df) * split_ratio)
    train = df.iloc[:split_idx]
    val   = df.iloc[split_idx:]
    return train, val

sj_train, sj_val = time_split(train_sj)
iq_train, iq_val = time_split(train_iq)

print("SJ train:", sj_train.shape, "SJ val:", sj_val.shape)
print("IQ train:", iq_train.shape, "IQ val:", iq_val.shape)

SJ train: (748, 25) SJ val: (188, 25)
IQ train: (416, 25) IQ val: (104, 25)


# Step 9: Prepare features & target (Model-ready)

In [ ]:
drop_cols = ["city", "week_start_date"]

features = [c for c in sj_train.columns if c not in drop_cols + ["total_cases"]]

X_sj_train = sj_train[features]
y_sj_train = sj_train["total_cases"]

X_sj_val = sj_val[features]
y_sj_val = sj_val["total_cases"]

X_iq_train = iq_train[features]
y_iq_train = iq_train["total_cases"]

X_iq_val = iq_val[features]
y_iq_val = iq_val["total_cases"]

print("Number of features:", len(features))

Number of features: 22


# Step 10: Train Polynomial Regression model


Model 1: Polynomial Regression (San Juan first)

In [ ]:
#start with degree = 2
#evaluate with MAE
#later we can try degree 3, 4 and compare

#import some libraries
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

In [ ]:
#Train Polynomial Regression (San Juan, degree = 2)
poly = PolynomialFeatures(degree=2, include_bias=False)

X_sj_train_poly = poly.fit_transform(X_sj_train)
X_sj_val_poly   = poly.transform(X_sj_val)

model = LinearRegression()
model.fit(X_sj_train_poly, y_sj_train)

y_sj_val_pred = model.predict(X_sj_val_poly)

mae_sj = mean_absolute_error(y_sj_val, y_sj_val_pred)
mae_sj

53.5233800659511

In [ ]:
#Train Polynomial Regression for Iquitos (degree = 2)
poly_iq = PolynomialFeatures(degree=2, include_bias=False)

X_iq_train_poly = poly_iq.fit_transform(X_iq_train)
X_iq_val_poly   = poly_iq.transform(X_iq_val)

model_iq = LinearRegression()
model_iq.fit(X_iq_train_poly, y_iq_train)

y_iq_val_pred = model_iq.predict(X_iq_val_poly)

mae_iq = mean_absolute_error(y_iq_val, y_iq_val_pred)
mae_iq

15.41449926693279

for k = 2

Iquitos is much easier to predict

San Juan has higher variance / complexity

# Step 11: Choose and justify polynomial degree (k)

In [ ]:
# here we try for k=2, k=3, k=4 to find the minimum polynomial degree

from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error

def eval_polynomial_degrees(X_train, y_train, X_val, y_val, degrees):
    results = {}
    for d in degrees:
        poly = PolynomialFeatures(degree=d, include_bias=False)

        X_train_poly = poly.fit_transform(X_train)
        X_val_poly   = poly.transform(X_val)

        model = LinearRegression()
        model.fit(X_train_poly, y_train)

        y_pred = model.predict(X_val_poly)
        mae = mean_absolute_error(y_val, y_pred)

        results[d] = mae
    return results

degrees = [2, 3, 4]

sj_results = eval_polynomial_degrees(
    X_sj_train, y_sj_train,
    X_sj_val, y_sj_val,
    degrees
)

iq_results = eval_polynomial_degrees(
    X_iq_train, y_iq_train,
    X_iq_val, y_iq_val,
    degrees
)

print("San Juan MAE:", sj_results)
print("Iquitos MAE:", iq_results)

San Juan MAE: {2: 53.5233800659511, 3: 303.76584742184076, 4: 646.9300720574139}
Iquitos MAE: {2: 15.41449926693279, 3: 46.05103810797021, 4: 55.958047389899704}


Results Summary

San Juan (MAE):

* Degree 2 → 53.52  [1st] (best)

* Degree 3 → 303.77 [2nd]

* Degree 4 → 646.93 [3rd]

Iquitos (MAE):

* Degree 2 → 15.41 [1st] (best)

* Degree 3 → 46.05 [2nd]

* Degree 4 → 55.96 [3rd]

In Conclusion:

* Higher-degree polynomials severely overfit

* Error explodes as degree increases

Degree = 2 balances:

* model capacity

* data size

* generalization

# Step 13 : Select ML model from literature
Import Gradient Boosting (Model 2)


In [ ]:
#importing libraries
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.metrics import mean_absolute_error

In [ ]:
#Train Gradient Boosting (San Juan)
gbr_sj = GradientBoostingRegressor(
    n_estimators=300,
    learning_rate=0.05,
    max_depth=3,
    random_state=42
)

gbr_sj.fit(X_sj_train, y_sj_train)

y_sj_val_pred = gbr_sj.predict(X_sj_val)

mae_sj_gbr = mean_absolute_error(y_sj_val, y_sj_val_pred)
mae_sj_gbr

18.30881136189219

* Polynomial k = 2 (best): ~53.52

* Gradient Boosting: ~18.31

* Massive MAE drop

In [ ]:
#Train Gradient Boosting (Iquitos)
gbr_iq = GradientBoostingRegressor(
    n_estimators=300,
    learning_rate=0.05,
    max_depth=3,
    random_state=42
)

gbr_iq.fit(X_iq_train, y_iq_train)

y_iq_val_pred = gbr_iq.predict(X_iq_val)

mae_iq_gbr = mean_absolute_error(y_iq_val, y_iq_val_pred)
mae_iq_gbr

7.506895238149951

| City     | Polynomial (k=2) | Gradient Boosting |
| -------- | ---------------- | ----------------- |
| San Juan | 53.52            | **18.31**         |
| Iquitos  | 15.41            | **7.51**          |

Gradient Boosting better for both cities

# Step 18 – Analyze feature importance

In [ ]:
#Feature importance for San Juan
import pandas as pd

fi_sj = pd.Series(
    gbr_sj.feature_importances_,
    index=features
).sort_values(ascending=False)

fi_sj.head(10)

,0
year,0.524293
weekofyear,0.174743
reanalysis_specific_humidity_g_per_kg,0.067774
reanalysis_dew_point_temp_k,0.046581
station_max_temp_c,0.029606
ndvi_se,0.023846
reanalysis_max_air_temp_k,0.019949
station_precip_mm,0.018835
reanalysis_avg_temp_k,0.016050
reanalysis_precip_amt_kg_per_m2,0.013113


In [ ]:
#Feature importance for Iquitos
fi_iq = pd.Series(
    gbr_iq.feature_importances_,
    index=features
).sort_values(ascending=False)

fi_iq.head(10)

,0
year,0.240333
weekofyear,0.189836
reanalysis_specific_humidity_g_per_kg,0.102156
reanalysis_avg_temp_k,0.059658
reanalysis_tdtr_k,0.055757
ndvi_sw,0.053503
reanalysis_dew_point_temp_k,0.047260
reanalysis_relative_humidity_percent,0.036898
reanalysis_precip_amt_kg_per_m2,0.031567
ndvi_se,0.029587


Feature importance analysis using the Gradient Boosting model indicates that temporal variables **(year and week of year)** are the most influential predictors for dengue incidence in both San Juan and Iquitos, highlighting strong seasonal patterns. Additionally, humidity-related variables such as specific humidity and dew point temperature consistently ranked among the top predictors, followed by temperature, precipitation, and vegetation indices (NDVI). These findings are consistent with prior studies emphasizing the role of eco-climatic conditions in dengue transmission.

# Step 20 : Build time-ahead forecasting models
Step 21 : Predict 1, 2, 3+ weeks ahead [DONE]

Step 22 : Evaluate MAE for each horizon [DONE]

Step 23 : Select best prediction horizon [DONE]


In [ ]:
from sklearn.metrics import mean_absolute_error
from sklearn.ensemble import GradientBoostingRegressor

def make_future_target(df, horizon):
    df = df.sort_values("week_start_date").copy()
    df["future_cases"] = df["total_cases"].shift(-horizon)
    return df.dropna(subset=["future_cases"])

def time_split(df, ratio=0.8):
    split = int(len(df) * ratio)
    return df.iloc[:split], df.iloc[split:]

def eval_horizons(df, horizons):
    results = {}
    for h in horizons:
        df_h = make_future_target(df, h)
        tr, va = time_split(df_h)

        Xtr, ytr = tr[features], tr["future_cases"]
        Xva, yva = va[features], va["future_cases"]

        model = GradientBoostingRegressor(
            n_estimators=300,
            learning_rate=0.05,
            max_depth=3,
            random_state=42
        )
        model.fit(Xtr, ytr)
        pred = model.predict(Xva)

        results[h] = mean_absolute_error(yva, pred)
    return results

horizons = [1, 2, 3, 4]

sj_horizon_mae = eval_horizons(train_sj, horizons)
iq_horizon_mae = eval_horizons(train_iq, horizons)

print("San Juan MAE by horizon:", sj_horizon_mae)
print("Iquitos MAE by horizon:", iq_horizon_mae)

San Juan MAE by horizon: {1: 17.54205194061072, 2: 17.822632401092207, 3: 18.245021248614744, 4: 18.71591723310891}
Iquitos MAE by horizon: {1: 7.56571169794041, 2: 9.147706482666724, 3: 8.059639310604638, 4: 11.170070946585248}


Results

San Juan (MAE):

* 1 week → 17.54 [1st] (best)

* 2 weeks → 17.82 [2nd]

* 3 weeks → 18.25 [3rd]

* 4 weeks → 18.72 [4th]

Iquitos (MAE):

* 1 week → 7.57 [1st] (best)

* 2 weeks → 9.15 [2nd]

* 3 weeks → 8.06 [3rd]

* 4 weeks → 11.17 [4th]

Error increases as horizon grows

1-week-ahead forecasting is most reliable

# STEP 24 — Train best model on FULL training data (weekly prediction)

In [ ]:
# Prepare full training sets (weekly prediction)
X_sj_full = train_sj[features]
y_sj_full = train_sj["total_cases"]

X_iq_full = train_iq[features]
y_iq_full = train_iq["total_cases"]

# Train final models
final_gbr_sj = GradientBoostingRegressor(
    n_estimators=300,
    learning_rate=0.05,
    max_depth=3,
    random_state=42
)
final_gbr_iq = GradientBoostingRegressor(
    n_estimators=300,
    learning_rate=0.05,
    max_depth=3,
    random_state=42
)

final_gbr_sj.fit(X_sj_full, y_sj_full)
final_gbr_iq.fit(X_iq_full, y_iq_full)

GradientBoostingRegressor(learning_rate=0.05, n_estimators=300, random_state=42)

In [ ]:
# Copy test data
test_df = X_test.copy()

# Identify numeric columns (same as training features)
numeric_cols = features

# Apply per-city mean imputation on TEST data
test_df[numeric_cols] = (
    test_df
    .groupby("city")[numeric_cols]
    .transform(lambda x: x.fillna(x.mean()))
)

# Sanity check
test_df[numeric_cols].isna().sum().sum()

np.int64(0)

#Step 25: Generate test predictions

In [ ]:
# Split test data by city
test_sj = test_df[test_df["city"] == "sj"].copy()
test_iq = test_df[test_df["city"] == "iq"].copy()

X_test_sj = test_sj[features]
X_test_iq = test_iq[features]

# Predict
test_sj["total_cases"] = final_gbr_sj.predict(X_test_sj)
test_iq["total_cases"] = final_gbr_iq.predict(X_test_iq)

# Combine into submission format
submission_df = pd.concat([test_sj, test_iq])[
    ["city", "year", "weekofyear", "total_cases"]
].sort_values(["city", "year", "weekofyear"])

submission_df.head()

,city,year,weekofyear,total_cases
260,iq,2010,26,3.306632
261,iq,2010,27,0.895914
262,iq,2010,28,5.208333
263,iq,2010,29,0.747746
264,iq,2010,30,1.524470


In [ ]:
submission_df.to_csv("submission.csv", index=False)

In [ ]:
#comparing two files for submission
import pandas as pd

submission_format = pd.read_csv("/content/submission_format.csv")
submission = pd.read_csv("/content/submission.csv")

print("Format shape:", submission_format.shape)
print("Submission shape:", submission.shape)

# Compare ID columns
ids_match = (
    submission_format[["city","year","weekofyear"]]
    .equals(submission[["city","year","weekofyear"]])
)

ids_match

Format shape: (416, 4)
Submission shape: (416, 4)


False

In [ ]:
import pandas as pd

submission_format = pd.read_csv("/content/submission_format.csv")
submission = pd.read_csv("/content/submission.csv")

In [ ]:
predictions = submission[["city", "year", "weekofyear", "total_cases"]]

In [ ]:
fixed_submission = submission_format.drop(columns=["total_cases"]).merge(
    predictions,
    on=["city", "year", "weekofyear"],
    how="left"
)

In [ ]:
# Check shape
print(fixed_submission.shape)

# Check missing predictions
print("NaNs in total_cases:", fixed_submission["total_cases"].isna().sum())

# Check ID alignment
print(
    submission_format[["city","year","weekofyear"]]
    .equals(fixed_submission[["city","year","weekofyear"]])
)

(416, 4)
NaNs in total_cases: 0
True


In [ ]:
fixed_submission.to_csv("submission_fixed.csv", index=False)

In [ ]:
fixed_submission["total_cases"] = (
    fixed_submission["total_cases"]
    .round()
    .clip(lower=0)
    .astype(int)
)

In [ ]:
fixed_submission.dtypes

,0
city,object
year,int64
weekofyear,int64
total_cases,int64


In [ ]:
fixed_submission.to_csv("submission_final.csv", index=False)